src/v0/destinations/klaviyo/util.js

```diff
@@ -309,6 +332,413 @@ const batchSubscribeEvents = (subscribeRespList) => {
+  batchEventResponse.batchedRequest.unshift(...profilesRequests);
+};
+
+/**
+ * This function returns the list of profileReq which do not metadata common with subcriptionMetadataArray
+ * @param {*} profileReq
+ * @param {*} subscriptionMetadataArray
+ * @returns
+ */
+const getRemainingProfiles = (profileReq, subscriptionMetadataArray) => {
+  const subscriptionListJobIds = subscriptionMetadataArray.map((metadata) => metadata.jobId);
+  return profileReq.filter((profile) => !subscriptionListJobIds.includes(profile.metadata.jobId));
+};
+/**
+ * This function batches the requests. Alogorithm
+ * Batch events from Subscribe Resp List having same listId/groupId to be subscribed and  have their metadata array
+ * For this metadata array get all profileRequests and add them prior to batched Subscribe Request in the same batched Request
+ * Make another batched request for the remaning profile requests and another for all the event requests
+ * @param {*} subscribeRespList
+ * @param {*} profileRespList
+ * @param {*} eventRespList
+ * subscribeRespList = [
+ * { payload: {id:'list_id', profile: {}}, metadata:{} },
+ * { payload: {id:'list_id', profile: {}}, metadata:{} }
+ * ]
+ * profileRespList = [
+ * { payload: {}, metadata:{} },
+ * { payload: {}, metadata:{} }
+ * ]
+ *
+ */
+const batchSubscriptionRequestV2 = (subscribeRespList, profileRespList, destination) => {
+  const batchedResponseList = [];
+  let remainingProfileReq = profileRespList;
+  const subscriptionMetadataArrayForAll = [];
+  const subscribeEventGroups = groupSubscribeResponsesUsingListIdV2(subscribeRespList);
+  Object.keys(subscribeEventGroups).forEach((listId) => {
+    // eventChunks = [[e1,e2,e3,..batchSize],[e1,e2,e3,..batchSize]..]
+    const eventChunks = lodash.chunk(subscribeEventGroups[listId], MAX_BATCH_SIZE);
+    const batchedResponse = [];
+    eventChunks.forEach((chunk) => {
+      const batchEventResponse = generateBatchedSubscriptionRequest(chunk, destination);
+      const { metadata: subscriptionMetadataArray, batchedRequest } = batchEventResponse;
+      updateBatchEventResponseWithProfileRequests(
+        remainingProfileReq,
+        subscriptionMetadataArray,
+        batchEventResponse,
+      );
+      subscriptionMetadataArrayForAll.push(...subscriptionMetadataArray);
+      batchedResponse.push(
+        getSuccessRespEvents(batchedRequest, subscriptionMetadataArray, destination, true),
+      );
+    });
+    batchedResponseList.push(...batchedResponse);
+  });
+  const profiles = [];
+  remainingProfileReq = getRemainingProfiles(remainingProfileReq, subscriptionMetadataArrayForAll);
+
+  // push profiles for which there is no subscription
+  remainingProfileReq.forEach((input) => {
+    profiles.push(
+      getSuccessRespEvents(
+        buildRequest(input.payload, destination, CONFIG_CATEGORIES.IDENTIFYV2),
+        [input.metadata],
+        destination,
+      ),
+    );
+  });
+  return [...profiles, ...batchedResponseList];
```

Comment:
Check if we further break this function into sub functions take help of GPT

Reply:
done